services/horizon/internal/expingest/ledger_entry_change_cache.go

```diff
@@ -0,0 +1,314 @@
+func (c *LedgerEntryChangeCache) addCreatedChange(change io.Change) error {
+	ledgerKeyString, err := change.Post.LedgerKey().MarshalBinaryBase64()
+	if err != nil {
+		return errors.Wrap(err, "Error MarshalBinaryBase64")
+	}
+
+	entryChange, exist := c.cache[ledgerKeyString]
+	if exist {
+		switch entryChange.Type {
+		case xdr.LedgerEntryChangeTypeLedgerEntryCreated:
+			return verify.NewStateError(errors.Errorf(
+				"Creating an entry that already exist. Ledger key = %s",
+				ledgerKeyString,
+			))
+		case xdr.LedgerEntryChangeTypeLedgerEntryUpdated:
+			return verify.NewStateError(errors.Errorf(
+				"Creating an entry that already exist. Ledger key = %s",
+				ledgerKeyString,
+			))
+		case xdr.LedgerEntryChangeTypeLedgerEntryRemoved:
+			// If existing type is removed it means that this entry does exist
+			// in a DB so we update entry change.
+			c.cache[ledgerKeyString] = xdr.LedgerEntryChange{
+				Type:    xdr.LedgerEntryChangeTypeLedgerEntryUpdated,
+				Updated: change.Post,
+			}
+		default:
+			return errors.Errorf("Unknown LedgerEntryChangeType: %d", entryChange.Type)
+		}
+	} else {
+		c.cache[ledgerKeyString] = xdr.LedgerEntryChange{
+			Type:    xdr.LedgerEntryChangeTypeLedgerEntryCreated,
+			Created: change.Post,
+		}
+	}
+
+	return nil
+}
+
+// addUpdatedChange adds a change to the cache, but returns an error if update
+// change is unexpected.
+func (c *LedgerEntryChangeCache) addUpdatedChange(change io.Change) error {
+	ledgerKeyString, err := change.Post.LedgerKey().MarshalBinaryBase64()
+	if err != nil {
+		return errors.Wrap(err, "Error MarshalBinaryBase64")
+	}
+
+	entryChange, exist := c.cache[ledgerKeyString]
+	if exist {
+		switch entryChange.Type {
+		case xdr.LedgerEntryChangeTypeLedgerEntryCreated:
+			// If existing type is created it means that this entry does not
+			// exist in a DB so we update entry change.
+			c.cache[ledgerKeyString] = xdr.LedgerEntryChange{
+				Type:    xdr.LedgerEntryChangeTypeLedgerEntryCreated,
+				Created: change.Post,
+			}
+		case xdr.LedgerEntryChangeTypeLedgerEntryUpdated:
+			c.cache[ledgerKeyString] = xdr.LedgerEntryChange{
+				Type:    xdr.LedgerEntryChangeTypeLedgerEntryUpdated,
+				Updated: change.Post,
+			}
+		case xdr.LedgerEntryChangeTypeLedgerEntryRemoved:
+			return verify.NewStateError(errors.Errorf(
+				"Updating an entry that was previously removed. Ledger key = %s",
+				ledgerKeyString,
+			))
+		default:
+			return errors.Errorf("Unknown LedgerEntryChangeType: %d", entryChange.Type)
+		}
+	} else {
+		c.cache[ledgerKeyString] = xdr.LedgerEntryChange{
+			Type:    xdr.LedgerEntryChangeTypeLedgerEntryUpdated,
+			Created: change.Post,
+		}
+	}
+
+	return nil
+}
+
+// addRemovedChange adds a change to the cache, but returns an error if remove
+// change is unexpected.
+func (c *LedgerEntryChangeCache) addRemovedChange(change io.Change) error {
+	ledgerKey := change.Pre.LedgerKey()
+	ledgerKeyString, err := ledgerKey.MarshalBinaryBase64()
```

Comment:
Can combine these lines as you did for the other funcs above:
```suggestion
ledgerKeyString, err := change.Pre.LedgerKey().MarshalBinaryBase64()
```

Reply:
`ledgerKey` is used twice below this line so worth keeping it in a variable.

Reply:
Actually the final code looks like in your first comment because ledgerKey isn't needed anymore. 😄